cpp/src/strings/capitalize.cu

```diff
@@ -0,0 +1,379 @@
+       int32_t bytes = 0;
+      
+       for( auto itr = d_str.begin(); itr != d_str.end(); ++itr ) {
+         auto the_chr = *itr;
+
+         auto pair_char_info = get_char_info(the_chr);
+         detail::character_flags_table_type flag = pair_char_info.second;
+         
+         if( (itr == d_str.begin()) ? IS_LOWER(flag) : IS_UPPER(flag) )
+           the_chr = convert_char(pair_char_info);
+         bytes += detail::bytes_in_char_utf8(the_chr);
+       }
+       return bytes;
+    }
+  };
+
+  //functor for executing string capitalization:
+  //(private inheritance to prevent polymorphic use,
+  // a requirement that came up in code review)
+  //
+  struct execute_capitalize: private execute_base
+  {
+    execute_capitalize(column_device_view const d_column,
+                       character_flags_table_type const* d_flags,
+                       character_cases_table_type const* d_case_table,
+                       int32_t const* d_offsets,
+                       char* d_chars):
+      execute_base(d_column, d_flags, d_case_table, d_offsets, d_chars)
+    {
+    }
+    
+    __device__
+    int32_t operator()(size_type idx) {
+      if( get_column().is_null(idx) )
+        return 0; // null string
+      
+      string_view d_str = get_column().template element<string_view>(idx);
+      char* d_buffer = get_chars() + get_offsets()[idx];
+      
+      for( auto itr = d_str.begin(); itr != d_str.end(); ++itr ) {
+        auto the_chr = *itr;
+
+        auto pair_char_info = get_char_info(the_chr); 
+        detail::character_flags_table_type flag = pair_char_info.second;
+
+        if( (itr == d_str.begin()) ? IS_LOWER(flag) : IS_UPPER(flag) )
+          the_chr = convert_char(pair_char_info);
+        d_buffer += detail::from_char_utf8(the_chr, d_buffer);
+      }
+      return 0;
+    }
+  };
+
+
+  //functor for probing string title-ization
+  //requirements:
+  //(private inheritance to prevent polymorphic use,
+  // a requirement that came up in code review)
+  //
+  struct probe_title: private probe_base
+  {
+    probe_title(column_device_view const d_column,
+                     character_flags_table_type const* d_flags,
+                     character_cases_table_type const* d_case_table):
+      probe_base(d_column, d_flags, d_case_table)
+    {  
+    }
+    
+     __device__
+     int32_t operator()(size_type idx) const {
+       if( get_column().is_null(idx) )
+         return 0; // null string
+      
+       string_view d_str = get_column().template element<string_view>(idx);
+       int32_t bytes = 0;
+
+       bool bcapnext = true;
+       for( auto itr = d_str.begin(); itr != d_str.end(); ++itr ) {
+         auto the_chr = *itr;
+
+         auto pair_char_info = get_char_info(the_chr);
+         
+         uint32_t code_point = pair_char_info.first;
+         detail::character_flags_table_type flag = pair_char_info.second;
+
+         if( !IS_ALPHA(flag) ) {
```

Comment:
I recommend refactoring this to match more closely to the execute code.
```
        if( !IS_ALPHA(flag) )
          bcapnext = true;
        else
          {
            if( bcapnext ? IS_LOWER(flag) : IS_UPPER(flag) )
              the_chr = convert_char(pair_char_info);
            bcapnext = false;
          }

        bytes += detail::bytes_in_char_utf8(the_chr);
```

Reply:
Done.